scalyr_agent/log_processing.py

```diff
@@ -750,6 +738,49 @@ def readline(self, current_time=None):
                     result.timestamp = timestamp
                     if attrs:
                         result.attrs = attrs
+                if self.__merge_json_parsed_lines:
+                    while result.line and not result.line.endswith(b"\n"):
+                        # Docker splits log lines at 16KB, we read lines until we hit a newline character and join them
+                        # so we can enforce our own max line length correctly.
+                        next_line = self.__read_extended_line(current_time)
+                        if len(next_line) == 0:
+                            # If we fail to read a line ending in a newline, we want to seek back to before we started
+                            # building this merged line and hopefully get all the bits next time around. We have a timer
+                            # that uses the line_completion_wait_time configuration used in LineMatchers to output the
+                            # line if we take too long waiting for a newline.
+                            if self.__merge_json_line_time is None:
+                                self.__merge_json_line_time = current_time
+
+                            if (
+                                current_time - self.__merge_json_line_time
+                                < self.__line_completion_wait_time
+                            ):
+                                self.__buffer.seek(original_buffer_position)
+                                self.__position = self.__determine_mark_position(
+                                    self.__buffer.tell()
+                                )
+                                log.log(
+                                    scalyr_logging.DEBUG_LEVEL_3,
+                                    "Incomplete merged line found in file %s, will reattempt reading.",
+                                    self.__path,
+                                )
+                                return LogLine(line=b"")
+                            else:
+                                self.__merge_json_line_time = None
+                                log.log(
+                                    scalyr_logging.DEBUG_LEVEL_3,
+                                    "Reached max wait time for incomplete merged line in file %s, emitting line as-is.",
+                                    self.__path,
+                                )
+                                break
+                        next_attrs = scalyr_util.json_decode(
+                            next_line.decode("utf-8", "replace")
+                        )
+                        line = next_attrs.pop(self.__json_log_key, None)
+                        if line is None:
```

Comment:
Why would `line` be `None` here?  Can you add in documentation capturing what this case represents.

Reply:
Actually I was missing a warning log for this case that appears when handling this elsewhere in the code, I think it's message describes this well enough (we didn't find the expected message field in the parsed JSON).

As well, when not merging lines and we run into a json line with no message we output the unparsed line. This isn't really something we can do mid-merge unless we tack it on to the end of the line so far, which is how I'll handle this for now so we don't lose data.